custom-plugin-decorators/sorting/sort-props-required.js

```diff
@@ -0,0 +1,33 @@
+module.exports = SortPropertiesRequiredFirst
+
+function SortPropertiesRequiredFirst() {
+  console.log("re-ordering properties: required first");
+  return {
+    Schema: {
+      leave(schema) {
+        if(schema.type == "object") {
+          const propList = Object.getOwnPropertyNames(schema.properties);
+          let newProps = {}
+
+          if(schema.required && schema.required.length > 0) {
+            const requiredList = schema.required;
+            // put the required items in first
+            requiredList.forEach((prop) => {
+              newProps[prop] = schema.properties[prop];
+            });
+
+            // now add anything that wasn't already added
+            propList.forEach((prop) => {
+              if(!newProps[prop]) {
+                newProps[prop] = schema.properties[prop];
+              }
+            });
+            schema.properties = newProps;
```

Comment:
```suggestion
          const { properties = {}, required = [] } = schema;
          for (const prop in properties) {
            if (!required.includes(prop)) {
              const p = properties[prop];
              // Delete and add the property back to the object if it's not required, so it will be at the end of the object:
              delete properties[prop];
              properties[prop] = p;
            }
          }
```

Reply:
This would work for this specific case, but I think as an overall pattern, the more verbose and readable approach of constructing a fresh array and replacing the old one is probably a more re-usable practice.